src/data/multiset/basic.lean

```diff
@@ -1746,6 +1746,22 @@ by { simp_rw [←filter_eq', eq_comm], congr }
 
 end
 
+lemma count_eq_card_filter_eq [decidable_eq α] (s : multiset α) (a : α) :
+  s.count a = (s.filter (eq a)).card :=
+begin
+  rw [count, countp_eq_card_filter],
+end
+
+@[simp] lemma map_count_true_eq_filter_card (s : multiset α) (p : α → Prop) [decidable_pred p] :
+  (s.map p).count true = (s.filter p).card :=
```

Comment:
This is weird: `classical.decidable` leaks here. 

The way this is written, Lean wants an instance `[decidable_eq Prop]` for `multiset.count`. Even in a file without any `open_locale classical`, etc, it finds this, through the linear order instance on `Prop`. This makes the decidability on the LHS different than on the RHS, so after the right rewrites, `rfl` still won't close it. (The proof should be `simp [count_eq_card_filter_eq, map_filter, card_map]`.) I suspect this might also stop the lemma from firing as a simp lemma.

The multiset of Props is a little weird to begin with. Would it make more sense to phrase this lemma in terms of `p : α → bool`?

Reply:
`filter` takes an argument of `α → Prop`, so I'm not sure how this would be changed to phrase it in terms of bools.

The point of this lemma is to prove
```
lemma filter_multiset_Ico_card_eq_of_periodic (n a : ℕ) (p : ℕ → Prop) [decidable_pred p]
  (pp : periodic p a) :
  (multiset.filter p (multiset.Ico n (n+a))).card = a.count p :=
```
In the upstream PR, where I definitely think it makes sense for `p` to be a predicate rather than a bool-valued function.

Reply:
Calling `simp [count_eq_card_filter_eq, map_filter, card_map]` from the beginning leaves us with an expression with `eq true ∘ p` in it, which I am surprised doesn't simplify to `p`. Are you saying this is due to an instance issue?

Reply:
Hmm, I see now what you are saying about the different instances. I guess this is unfortunate for someone trying to understand this proof, but as far as applying this lemma to my other things, it works fine, so I would prefer not to change it.

Reply:
> Calling `simp [count_eq_card_filter_eq, map_filter, card_map]` from the beginning leaves us with an expression with `eq true ∘ p` in it, which I am surprised doesn't simplify to `p`.

Right. Even if you do this:
```lean
@[simp] lemma eq_true_comp (P : α → Prop) : (eq true ∘ P) = P :=
funext $ λ x, by simp [eq_true]

@[simp] lemma map_count_true_eq_filter_card (s : multiset α) (p : α → Prop) [decidable_pred p] :
  (s.map p).count true = (s.filter p).card :=
begin
  simp [count_eq_card_filter_eq, map_filter, card_map, eq_true_comp],
end
```
you get the goal `⇑card (filter p s) = ⇑card (filter p s)` which is *not* a refl goal, because there are different decidable instances. This may actually change with Lean 3.38.0c, released four hours ago -- but they are legitimately different decidability instances.

There are a few issues mixed together here. `count` probably has too strong of a decidability requirement. You shouldn't need to decide whether any two props are equal, just whether a prop is equal to the thing you're trying to count: in your case, whether `p : Prop` is equal to `true`, which can be done with `decidable_pred P`. But the linear order on `Prop` is leaking `classical.dec_eq` . You sould be able to `#eval (({0, 0, 1} : multiset ℕ).map (> 0)).count true ` but you can't. This leaking makes your proof more complicated than it needs to be, and also more specialized: it will only apply to cases where the leaky classical gets used. (Which right now is probably all of them, but still.)